### DISCUSSION TITLE
Geant4 Installation Issue

### DISCUSSION THREAD

**Discussion by Ljoseph5454** [mctools/ncrystal#137] [Q&A] [answered]
Hi,

I am having an issue with implementing NCrystal in Geant4. First I followed the INSTALL file and successfully installed it and performed the test successfully. (I made sure to cmake with -DNCRYSTAL_ENABLE_GEANT4=ON). Then, I went into the folder with my Geant4 simulations and did the following: (Note that my file is called DMX instead of exampleapp as in the README)

1) I added #include "G4NCrystal/G4NCrystal.hh" to DMX.cc
2) In CMakeLists.txt I added the following commands: find_package(NCrystal REQUIRED COMPONENTS GEANT4BINDINGS) target_link_libraries(DMX NCrystal::G4NCrystal) Now I had the structure of what is given in the README as an example
I put the full cmake file below
3) I ran: cmake -DDMXENV_GPS_USE=ON OK -DGeant4_DIR=/data/ljoseph/geant4.10.03.p03-build/geant4-10.3.3 -DNCrystal_DIR=/data/ljoseph/ncrystalinstall /data/ljoseph/Geant4/TN/thermal_neutron 
The first directory in the command is where my Geant is installed, second directory is where NCrystal is, and last directory is where my simulation is. However, this threw an error

The error was: Make Error at /data/ljoseph/ncrystalinstall/lib64/cmake/NCrystal/NCrystalConfig.cmake:118 (if):
  if given arguments:

    "GEANT4BINDINGS" "IN_LIST" "NCrystal_FIND_COMPONENTS"

  Unknown arguments specified
Call Stack (most recent call first):
  CMakeLists.txt:7 (find_package)

The error seems to imply that the config file (which I have not messed with) has an if command: if ( "GEANT4BINDINGS" IN_LIST NCrystal_FIND_COMPONENTS ) that has unrecognized arguments. I must be doing something wrong as I doubt it is an issue with the NCrystal file. However, I did grep for NCrystal_FIND_COMPONENTS and it was only referenced in NCrystalConfig.cmake so it almost seems undefined as the error suggests. Below I have pasted my full CMakeLists file. Any help is greatly appreciated! We are not sure where to go from here.

#----------------------------------------------------------------------------
# Setup the project
cmake_minimum_required(VERSION 2.6 FATAL_ERROR)
project(underground_physics)

#Logan Added
find_package(NCrystal REQUIRED COMPONENTS GEANT4BINDINGS)

option(WITH_GEANT4_UIVIS "Build example with Geant4 UI and Vis drivers" ON)
if(WITH_GEANT4_UIVIS)
  find_package(Geant4 REQUIRED ui_all vis_all)
else()
  find_package(Geant4 REQUIRED)
endif()

include(${Geant4_USE_FILE})

if("$ENV{DMXENV_GPS_USE}" AND NOT DEFINED DMXENV_GPS_USE)
  set(DMXENV_GPS_USE_DEFAULT ON)
else()
  set(DMXENV_GPS_USE_DEFAULT ${DMXENV_GPS_USE})
endif()

option(DMXENV_GPS_USE "Build DMX with support for GPS" ${DMXENV_GPS_USE_DEFAULT})
mark_as_advanced(DMXENV_GPS_USE)

if(DMXENV_GPS_USE)
  add_definitions(-DDMXENV_GPS_USE)
endif()

include_directories(${PROJECT_SOURCE_DIR}/include 
                    ${Geant4_INCLUDE_DIR})
file(GLOB sources ${PROJECT_SOURCE_DIR}/src/*.cc)
file(GLOB headers ${PROJECT_SOURCE_DIR}/include/*.hh)

add_executable(DMX DMX.cc ${sources} ${headers})
target_link_libraries(DMX ${Geant4_LIBRARIES})

#Logan Added
target_link_libraries(DMX NCrystal::G4NCrystal)

set(underground_physics_SCRIPTS
    alpha_1000.mac alpha_low.mac alpha.mac ambe_spectrum.mac gamma_1000.mac gamma_2000.mac gamma_low.mac gamma.mac initInter.mac neutron.mac stt.mac stt.out stt_pmt_ref.out stt_ref.out wired.mac Th232.mac U238_low.mac U238_up.mac Rn222.mac
  )

foreach(_script ${underground_physics_SCRIPTS})
  configure_file(
    ${PROJECT_SOURCE_DIR}/${_script}
    ${PROJECT_BINARY_DIR}/${_script}
    COPYONLY
    )
endforeach()

add_custom_target(underground_physics DEPENDS DMX)

install(TARGETS DMX DESTINATION bin)

**Comment by tkittel** [comment]
Thanks for the report. What CMake version are you using?

**Reply by Ljoseph5454** [reply]
Sorry for the late response, but I am using cmake version 3.10.2

**Comment by tkittel** [comment]
I think I understand your issue now. I will write more details later after a few more investigattions, but for now if you change the version number in `cmake_minimum_required(VERSION 2.6 FATAL_ERROR)` from 2.6 to at least 3.3, I think there is a good chance it will work for you.

**Reply by Ljoseph5454** [reply]
Hi,  I deleted my old comment but this worked! However, now when I am trying to make the example geant4 code by doing "make -j 2 name_of_file", I get the following error

In file included from /data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:28:0,
                 from /data/ljoseph/ncrystalinstall/include/NCrystal/NCProcImpl.hh:24,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCManager.hh:29,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCrystal.hh:32,
                 from /data/ljoseph/Geant4/TN/NCrystalTest/DMX.cc:33:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh: In destructor \u2018NCrystal::ImmutableBuffer<LOCALBUF_MINSIZE, BUF_ALIGNMENT, TMetaData>::~ImmutableBuffer()\u2019:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh:154:20: error: \u2018is_trivially_copyable\u2019 is not a member of \u2018std\u2019
     static_assert( std::is_trivially_copyable<TMetaData>::value, "");
                    ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh:154:56: error: expected primary-expression before \u2018>\u2019 token
     static_assert( std::is_trivially_copyable<TMetaData>::value, "");
                                                        ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh:154:57: error: \u2018::value\u2019 has not been declared
     static_assert( std::is_trivially_copyable<TMetaData>::value, "");
                                                         ^
In file included from /data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:31:0,
                 from /data/ljoseph/ncrystalinstall/include/NCrystal/NCProcImpl.hh:24,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCManager.hh:29,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCrystal.hh:32,
                 from /data/ljoseph/Geant4/TN/NCrystalTest/DMX.cc:33:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh: In instantiation of \u2018class NCrystal::Variant<double, int>\u2019:
/usr/include/c++/4.8.2/type_traits:706:28:   required from \u2018struct std::__is_default_constructible_impl<NCrystal::Variant<double, int> >\u2019
/usr/include/c++/4.8.2/type_traits:116:12:   required from \u2018struct std::__and_<std::__not_<std::is_void<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)0> > >, std::__is_default_constructible_impl<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)0> > >\u2019
/usr/include/c++/4.8.2/type_traits:721:12:   required from \u2018struct std::__is_default_constructible_atom<NCrystal::Variant<double, int> >\u2019
/usr/include/c++/4.8.2/type_traits:742:12:   required from \u2018struct std::__is_default_constructible_safe<NCrystal::Variant<double, int>, false>\u2019
/usr/include/c++/4.8.2/type_traits:748:12:   required from \u2018struct std::is_default_constructible<NCrystal::Variant<double, int> >\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:116:67:   required from here
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:101:46: error: requested alignment is not an integer constant
     alignas(align_req) uint8_t m_data[MAXSIZE];
                                              ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh: At global scope:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:116:3: error: static assertion failed: 
   static_assert(std::is_default_constructible<Variant<double,int>>::value,"");
   ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh: In instantiation of \u2018class NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1>\u2019:
/usr/include/c++/4.8.2/type_traits:706:28:   required from \u2018struct std::__is_default_constructible_impl<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> >\u2019
/usr/include/c++/4.8.2/type_traits:116:12:   required from \u2018struct std::__and_<std::__not_<std::is_void<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> > >, std::__is_default_constructible_impl<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> > >\u2019
/usr/include/c++/4.8.2/type_traits:721:12:   required from \u2018struct std::__is_default_constructible_atom<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> >\u2019
/usr/include/c++/4.8.2/type_traits:742:12:   required from \u2018struct std::__is_default_constructible_safe<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1>, false>\u2019
/usr/include/c++/4.8.2/type_traits:748:12:   required from \u2018struct std::is_default_constructible<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> >\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:117:90:   required from here
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:101:46: error: requested alignment is not an integer constant
     alignas(align_req) uint8_t m_data[MAXSIZE];
                                              ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh: In instantiation of \u2018class NCrystal::Variant<NCrystal::CrystalAxis, NCrystal::HKLPoint, (NCrystal::VariantAllowEmpty)1>\u2019:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:378:13:   required from here
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:101:46: error: requested alignment is not an integer constant
In file included from /data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:28:0,
                 from /data/ljoseph/ncrystalinstall/include/NCrystal/NCProcImpl.hh:24,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCManager.hh:29,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCrystal.hh:32,
                 from /data/ljoseph/Geant4/TN/NCrystalTest/DMX.cc:33:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh: In instantiation of \u2018class NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId>\u2019:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCSmallVector.hh:267:77:   required from \u2018struct NCrystal::detail::SVTestCB<NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId> >::<anonymous>\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCSmallVector.hh:267:89:   required from \u2018class NCrystal::detail::SVTestCB<NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId> >\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCSmallVector.hh:275:50:   required from \u2018constexpr bool NCrystal::detail::SVUseFast() [with TValue = NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId>; NCrystal::SVMode MODE = (NCrystal::SVMode)2]\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCSmallVector.hh:52:100:   required from \u2018class NCrystal::SmallVector<NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId>, 7ul, (NCrystal::SVMode)2>\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:510:28:   required from here
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh:125:54: error: requested alignment is not an integer constant
     alignas(object_alignment) char m_data[object_size];
                                                      ^
make[3]: *** [CMakeFiles/DMX.dir/DMX.cc.o] Error 1
make[2]: *** [CMakeFiles/DMX.dir/all] Error 2
make[1]: *** [CMakeFiles/DMX.dir/rule] Error 2
make: *** [DMX] Error 2
(base) [ljoseph@twophoton build]$ ^C
(base) [ljoseph@twophoton build]$ ^C
(base) [ljoseph@twophoton build]$ gedit /data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh
(base) [ljoseph@twophoton build]$ cmake -DDMXENV_GPS_USE=ON -DGeant4_DIR=/data/ljoseph/geant4.10.03.p03-build/geant4-10.3.3 -DNCrystal_DIR=/data/ljoseph/ncrystalinstall /data/ljoseph/Geant4/TN/NCrystalTest
-- Configuring done
-- Generating done
-- Build files have been written to: /data/ljoseph/Geant4/TN/NCrystalTest/build
(base) [ljoseph@twophoton build]$ make -j 2 DMX
[ 50%] Building CXX object CMakeFiles/DMX.dir/DMX.cc.o
In file included from /data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:28:0,
                 from /data/ljoseph/ncrystalinstall/include/NCrystal/NCProcImpl.hh:24,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCManager.hh:29,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCrystal.hh:32,
                 from /data/ljoseph/Geant4/TN/NCrystalTest/DMX.cc:31:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh: In destructor \u2018NCrystal::ImmutableBuffer<LOCALBUF_MINSIZE, BUF_ALIGNMENT, TMetaData>::~ImmutableBuffer()\u2019:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh:156:20: error: \u2018is_trivially_copyable\u2019 is not a member of \u2018std\u2019
     static_assert( std::is_trivially_copyable<TMetaData>::value, "");
                    ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh:156:56: error: expected primary-expression before \u2018>\u2019 token
     static_assert( std::is_trivially_copyable<TMetaData>::value, "");
                                                        ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh:156:57: error: \u2018::value\u2019 has not been declared
     static_assert( std::is_trivially_copyable<TMetaData>::value, "");
                                                         ^
In file included from /data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:31:0,
                 from /data/ljoseph/ncrystalinstall/include/NCrystal/NCProcImpl.hh:24,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCManager.hh:29,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCrystal.hh:32,
                 from /data/ljoseph/Geant4/TN/NCrystalTest/DMX.cc:31:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh: In instantiation of \u2018class NCrystal::Variant<double, int>\u2019:
/usr/include/c++/4.8.2/type_traits:706:28:   required from \u2018struct std::__is_default_constructible_impl<NCrystal::Variant<double, int> >\u2019
/usr/include/c++/4.8.2/type_traits:116:12:   required from \u2018struct std::__and_<std::__not_<std::is_void<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)0> > >, std::__is_default_constructible_impl<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)0> > >\u2019
/usr/include/c++/4.8.2/type_traits:721:12:   required from \u2018struct std::__is_default_constructible_atom<NCrystal::Variant<double, int> >\u2019
/usr/include/c++/4.8.2/type_traits:742:12:   required from \u2018struct std::__is_default_constructible_safe<NCrystal::Variant<double, int>, false>\u2019
/usr/include/c++/4.8.2/type_traits:748:12:   required from \u2018struct std::is_default_constructible<NCrystal::Variant<double, int> >\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:116:67:   required from here
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:101:46: error: requested alignment is not an integer constant
     alignas(align_req) uint8_t m_data[MAXSIZE];
                                              ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh: At global scope:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:116:3: error: static assertion failed: 
   static_assert(std::is_default_constructible<Variant<double,int>>::value,"");
   ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh: In instantiation of \u2018class NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1>\u2019:
/usr/include/c++/4.8.2/type_traits:706:28:   required from \u2018struct std::__is_default_constructible_impl<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> >\u2019
/usr/include/c++/4.8.2/type_traits:116:12:   required from \u2018struct std::__and_<std::__not_<std::is_void<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> > >, std::__is_default_constructible_impl<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> > >\u2019
/usr/include/c++/4.8.2/type_traits:721:12:   required from \u2018struct std::__is_default_constructible_atom<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> >\u2019
/usr/include/c++/4.8.2/type_traits:742:12:   required from \u2018struct std::__is_default_constructible_safe<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1>, false>\u2019
/usr/include/c++/4.8.2/type_traits:748:12:   required from \u2018struct std::is_default_constructible<NCrystal::Variant<double, int, (NCrystal::VariantAllowEmpty)1> >\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:117:90:   required from here
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:101:46: error: requested alignment is not an integer constant
     alignas(align_req) uint8_t m_data[MAXSIZE];
                                              ^
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh: In instantiation of \u2018class NCrystal::Variant<NCrystal::CrystalAxis, NCrystal::HKLPoint, (NCrystal::VariantAllowEmpty)1>\u2019:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:378:13:   required from here
/data/ljoseph/ncrystalinstall/include/NCrystal/NCVariant.hh:101:46: error: requested alignment is not an integer constant
In file included from /data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:28:0,
                 from /data/ljoseph/ncrystalinstall/include/NCrystal/NCProcImpl.hh:24,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCManager.hh:29,
                 from /data/ljoseph/ncrystalinstall/include/G4NCrystal/G4NCrystal.hh:32,
                 from /data/ljoseph/Geant4/TN/NCrystalTest/DMX.cc:31:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh: In instantiation of \u2018class NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId>\u2019:
/data/ljoseph/ncrystalinstall/include/NCrystal/NCSmallVector.hh:267:77:   required from \u2018struct NCrystal::detail::SVTestCB<NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId> >::<anonymous>\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCSmallVector.hh:267:89:   required from \u2018class NCrystal::detail::SVTestCB<NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId> >\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCSmallVector.hh:275:50:   required from \u2018constexpr bool NCrystal::detail::SVUseFast() [with TValue = NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId>; NCrystal::SVMode MODE = (NCrystal::SVMode)2]\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCSmallVector.hh:52:100:   required from \u2018class NCrystal::SmallVector<NCrystal::ImmutableBuffer<24ul, 8ul, NCrystal::Cfg::detail::VarId>, 7ul, (NCrystal::SVMode)2>\u2019
/data/ljoseph/ncrystalinstall/include/NCrystal/NCTypes.hh:510:28:   required from here
/data/ljoseph/ncrystalinstall/include/NCrystal/NCImmutBuf.hh:127:54: error: requested alignment is not an integer constant
     alignas(object_alignment) char m_data[object_size];
                                                      ^
make[3]: *** [CMakeFiles/DMX.dir/DMX.cc.o] Error 1
make[2]: *** [CMakeFiles/DMX.dir/all] Error 2
make[1]: *** [CMakeFiles/DMX.dir/rule] Error 2
make: *** [DMX] Error 2

**Reply by Ljoseph5454** [reply]
If easier, here is a screen shot of the first bit of errors 
![image](https://github.com/mctools/ncrystal/assets/113142116/01aa3b69-77ec-458b-a00f-60615c6ee5dd)

**Comment by tkittel** [comment]
Looks to me like your compiler is not being configured for C++11. Can you provide a tar-ball somewhere with a complete (and minimal!) example of your .cpp and CMakeLists.txt files, and I can see if I can reproduce it?

**Reply by Ljoseph5454** [reply]
Hopefully this is what you are looking for. I had tried NCrystal in 2 ways. The first way (NCrystalTest.tar) is me using the example code on this GitHub (I copied my CMakeLists from my other test so its probably more complicated than needed) so this one is simpler. The other way (thermal_neutron_test.tar) is the directory that I care about and has a lot more stuff. My only warning here is that the CMakeLists might be looking for some macro files that are not in the .tar (since I got rid of unnecessary stuff), so that line might have to be deleted. Apologies if I am missing anything, this is my first time doing this. 
[NCrystalTest.tar.gz](https://github.com/mctools/ncrystal/files/12244960/NCrystalTest.tar.gz)
[thermal_neutron_test.tar.gz](https://github.com/mctools/ncrystal/files/12244961/thermal_neutron_test.tar.gz)

**Reply by tkittel** [reply]
To be honest I was hoping for a more minimal example, without any extraneous stuff not related to the issue at hand. But in any case, I was wondering if perhaps adding a `set(CMAKE_CXX_STANDARD 11)` to your CMake code would not help (right after the `project(..)` statement for instance).

And to be clear: These issues arise because you are using the old-school CMake 2.x style of CMake code. The recommended way is simply to add a dependency on one of the exported targets from NCrystal.

**Reply by Ljoseph5454** [reply]
Im sorry, I thought maybe the Geant4 example on this Github would be good enough. Just to be clear, did you look at the NCrystalTest.tar.gz? I did shave down a lot of unnecessary stuff though, so maybe try this tar below (the one called NCrystalTestSimple.tar.gz). Hopefully this one suffices. I did notice a few new things. I changed project(name) to project(name LANGUAGES CXX) (I also tried project(name LANGUAGES CXX C), as mentioned in the README, and the example file actually maked properly, but running it gave a segmentation fault. I tried this on my other projects and it just gave other errors. However, It doesn't seem like that addition is necessary. Removing it gives the original error I sent you. Also, it might be good to be clear that I am running the right CMAKE command. In my simulation build folder I run cmake -DGeant4_DIR=/data/ljoseph/geant4.10.03.p03-build/geant4-10.3.3 -DNCrystal_DIR=/data/ljoseph/ncrystalinstall /data/ljoseph/Geant4/TN/NCrystalTest where the last path is the path to my simulation. Then I run make -j2 DMX (DMX is the name of my main .cc file). This is correct right? I appreciate all your help so far and I apologize for my struggles, I am still learning.

Another thing which might be very helpful if you would be willing to share a CMakeLists file which is written for more modern CMake versions. I inherited these files from an older project and none of us on the team are well-versed enough in CMake to write an updated one, so seeing a CMakeLists example that works would be helpful if you are willing to share one. 

[NCrystalTestSimple.tar.gz](https://github.com/mctools/ncrystal/files/12254898/NCrystalTestSimple.tar.gz)

**Reply by tkittel** [reply]
So trying your NCrystalTestSimple.tar.gz files, they actually work fine out of the box for me, although my machine might have a different setup that yours. One thing you can try, is to try to get the latest NCrystal instead: v3.7.1 came out yesterday, and include a few fixes inspired by your use-case. In particular, I believe I have fixed the original CMake error, and maybe the error about is_trivially_copyable as well (not 100% sure, since I never had that error in the first place on any of our tested systems).

Although it won't help you right now, I should mention that my plans for the fall include revisiting the NCrystal-Geant4 hooks, and hopefully will also end up in better documentation, recipes, examples, ...

Cheers,
Thomas

**Reply by Ljoseph5454** [reply]
Thanks for all your help! It seems like this patch is better. While I am still having some errors in compiling my code (which works in base Geant4), the example code on this github is now fully compiling. However, running it gives a segmentation fault and I discovered that it is faulting at the line which creates the physics lists:
runManager->SetUserInitialization(G4PhysListFactory().GetReferencePhysList("QGSP_BIC_HP"));
 (and it is unhappy with the particle gun code). If it worked for you, what might you think the issue is there? Also where might I find an example of a working CMakeLists that is written in a more modern style?

Kindly,
Logan

**Reply by tkittel** [reply, marked as answer]
To answer the first half, I can only guess. Perhaps you have an incomplete geant4-datafile setup which (perhaps) only triggers with an HP physics list? Otherwise, I was using geant4-10.04.p03 to test, which differs from yours.

For the second question, note that I am currently working on better Geant4-NCrystal hooks which will also result in updated documentation. But as a quick example, here is a modern (aka "cmake3" or "target-based") cmake lists which works for me:

```
cmake_minimum_required(VERSION 3.10...3.24)
project(MyExampleProject LANGUAGES CXX)
find_package( NCrystal REQUIRED COMPONENTS GEANT4BINDINGS )
add_executable(exampleapp "${CMAKE_CURRENT_SOURCE_DIR}/exampleapp.cpp")
target_link_libraries( exampleapp NCrystal::G4NCrystal )
install( TARGETS exampleapp DESTINATION bin )
```
Note that you get the geant4 dependency (with link and compilation flags) injected automatically when linking against the `NCrystal::G4NCrystal` target.  I do not like to use a `include(${Geant4_USE_FILE})` statement, since it modifies the global behaviour for all targets, which is exactly what is frowned upon (for good reasons) in modern CMake code. However, admittedly, the usage of `include(${Geant4_USE_FILE})` might still be the recommended way mentioned in most official Geant4 documentation. Hopefully that will change eventually.

**Reply by Ljoseph5454** [reply]
Hi Thomas,

We finally got the example code to run! Thank you so much for your help I doubt we could have gotten it to work otherwise - we really appreciate it. My code is still not working but its clear that that is my issue now and not NCrystal.

**Reply by tkittel** [reply]
Good to hear that you at least managed to get past this hurdle! And hopefully usage of Geant4+NCrystal will become a lot more streamlined later this fall, with better documentation etc.

**Comment by tkittel** [comment]
And please also provide information about your system (in particular compiler version, how you installed it, etc.).

**Reply by Ljoseph5454** [reply]
Hi, here is a screenshot I took of stuff I thought was relevant 
![image](https://github.com/mctools/ncrystal/assets/113142116/5735e7d0-ae79-4f0b-8b7d-1841304dbbc9)

I am still new to linux and have gotten help from others in my group. As a result I do not know exactly how it was installed but if necessary I can search around for who knows. In summary though, gcc version is 12.2, python is 3.10.8, and the linux release is 7.9, I am also using python through conda

**Reply by tkittel** [reply]
So one thing that strikes me here is that your are using python through conda, but a compiler which is not provided by conda. Such a mix is most likely not ideal.

**Reply by Ljoseph5454** [reply]
Right, this is due to how the remote computer is setup, I am working on getting python downloaded so I can access it not through Conda.
